### PR TITLE
Initial Event Bus implementation & Event-based multiplayer updates

### PR DESCRIPTION
**Pull request by Azzurite**
This is *still* just another PR in preparation for the feature I actually want to add :D

In total, this is basically just a refactoring, there should be basically no observable changes, except where otherwise noted.

A [central event bus](https://gameprogrammingpatterns.com/event-queue.html) is added. It is based on standard kotlin classes that extend `Event`. This imo makes it very nice to work with since you can add any data with the event you like, immediately with no boilerplate code. You also can see all events just by telling your IDE to list all implementing classes. In addition, it was requested that it not be asynchronous/threaded for better debuggability, so you are only supposed to use it on the main render thread. This thread was chosen because when you listen for events, you mostly want to do something with the UI immediately based on the event. Otherwise I hope it's pretty simple & self-explanatory. 

(This "non-multithreaded requirement" also means I can't show off very nice coroutine features :D the whole `EventBus` file could be shrinked to 10 lines with added asynchronous support, at the expense of being able to step through from send->receive in the debugger)

The other part of this PR is a major refactor of the multiplayer logic to be centralized in their own classes in `com.unciv.logic.multiplayer` instead of being strewn around various `Screen`s. And instead of using the "throw away everything and rebuild based on the newest data"-approach that has mostly been used so far, it instead now uses events from the event bus to listen for changes and only update the specific components that need updating. 

(This is not really noticeable in the `MultiplayerScreen` because not much is happening there, but when you extend this approach to more performance-critical parts of the game, this will get us major performance benefits. We currently  "if something changed, throw away things and rebuild based on the newest data" *a lot* of things when a `WorldScreen.update` happens - if all those things would not need a "has updated" check and only update the small parts that actually change, the whole game would be lightning-fast)

More possibly interesting comments as code comments down below:

### CODE REVIEW

File: core/src/com/unciv/logic/multiplayer/OnlineMultiplayer.kt [hunk 1]
```diff
@@ -0,0 +1,284 @@
...
+        val saveFileName = if (gameName.isNullOrBlank()) gameId else gameName
+        var gamePreview: GameInfoPreview
+        var fileHandle: FileHandle
+        try {
+            gamePreview = OnlineMultiplayerGameSaver().tryDownloadGamePreview(gameId)
+            fileHandle = GameSaver.saveGame(gamePreview, saveFileName)
+        } catch (ex: FileNotFoundException) {
+            // Game is so old that a preview could not be found on dropbox lets try the real gameInfo instead
+            gamePreview = OnlineMultiplayerGameSaver().tryDownloadGame(gameId).asPreview()
+            fileHandle = GameSaver.saveGame(gamePreview, saveFileName)
```
[the remark this answers is not in the document]
Comment (SomeTroglodyte) [reply core/src/com/unciv/logic/multiplayer/OnlineMultiplayer.kt:129]: One of the open problems around here: No auto cleanup on "Unciv's" Dropbox.

[the remark this answers is not in the document]
Comment (xlenstra) [reply core/src/com/unciv/logic/multiplayer/OnlineMultiplayer.kt:129]: I manually remove a lot of them every time the dropbox fills up. Currently, all save games up to (but excluding) 2021-10-31 are removed from the dropbox

[the remark this answers is not in the document]
Comment (SomeTroglodyte) [reply core/src/com/unciv/logic/multiplayer/OnlineMultiplayer.kt:129]: Ah yes there was some python script and idea to make it a github action...

**Comment by itanasi** [comment]
This would have been a nice way to track which cities need to re-evaluate in my PR #6650

**Comment by SomeTroglodyte** [comment]
Six pages essentially a monologue so far... OK I'll **_read_** the beast.

> `open class Event`

OK, this is key. In fact, I _must_ find an excuse to disparage a 16-char oneliner.... ... .. . OK: Probability for Name conflicts with std libraries too high? Auto-Imports of Group, Array, Timer etc all run into problems that maybe the wrong one is chosen... Extra namespace "EventBus.Event" (need for qualifier can be import-tweaked away if needed - per-file)? "UncivEvent" (too long)?. "Efent" (:rofl:)? We'll probably just have to be careful. Oh, and please end all new files with a LF for prettier github rendering.

> `java.lang.ref.WeakReference`

There's no kotlin [equivalent](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.native.ref/-weak-reference/)? Even if it turns out to be just a typealias...

> `private fun <T: Event> receive`

This does not actually _receive_, right? It's more a listen to or register handler or somesuch. Oh, private, the meat is elsewhere. never mind - yes, the dox of the actual meat make it clear. (Besides, isn't the '[recieve](https://en.wiktionary.org/wiki/recieve)' typo obligatory by now?)

> `class MultiplayerGameUpdateErrored(...) : Event()`

'Errored' - uh, okay, can't think of a better name either. And - why is Event an open class instead of an interface anyway?  Would save Gigabytes of `()` all over the place...

At the UI stuff I started to skip stuff, trust will do.
I like the EventBus design and can't wait to see all our Stats updates riding it.
So...

**Review by SomeTroglodyte — APPROVED** [review]

**Comment by SomeTroglodyte** [comment]
> when being asked "How are you?"

Good :tm: for you. When asked that I always faint. :alien: :man_health_worker: 

> Name conflicts

We had precedences. Java math where using kotlin's consistently would be better, and one with just a tiny little more weight I forget? And for some Gdx classes I always think "does everybody on the project know _not_ to pick the first offer"? Not important.

> There's no kotlin [equivalent](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.native.ref/-weak-reference/)?

That one had a link. Duckduck instead of Studio. May be out of context, didn't really dive in.

**Comment by SomeTroglodyte** [comment]
Ah, out of context. Got it. :woman_facepalming:

**Comment by SomeTroglodyte** [comment]
> setSizeForced

Not really part of Event Bus, right?    .... Split it off?

File: android/build.gradle.kts [hunk 2]
```diff
@@ -127,4 +129,5 @@ dependencies {
     //   run `./gradlew build --scan` to see details
     implementation("androidx.core:core-ktx:1.6.0")
     implementation("androidx.work:work-runtime-ktx:2.6.0")
+    coreLibraryDesugaring("com.android.tools:desugar_jdk_libs:1.1.5")
```
[the remark this answers is not in the document]
Comment (touhidurrr) [reply android/build.gradle.kts:132]: I remember arguing over desugaring once. When Unciv wanted to bumb its minimum android sdk version. But it was shot down.

**Comment by SomeTroglodyte** [comment]
I see the merge was easy, yes. I reaffirm I think this will be a boon - we'll just have to get people to actually use it. Recalc stats, reassign population,...

Merge with just our two votes? Or _request_ a review from someone specific first?

**Comment by yairm210** [comment]
I think that once we start implementing things using this, we should add tests to validate previous behavior, and ensure that that behavior remains after refactor

**Comment by GGGuenni** [comment]
This has 11 `Call requires API level 26 (current min is 21)` errors. All related to java.time. Doesn't this result in unciv being unable to run on Lollipop, Marshmallow, and Nougat?

**Comment by touhidurrr** [comment]
Just testing Github's email capabilities.
Also what is this event bus for?
Will it trigger an event for every change in GameInfo so that we can handle
those events or something like that?
I mean what does it cover???

**Comment by touhidurrr** [comment]
Just testing Githubs email capabilities.
Also what is this event bus for?
Will it trigger an event for every change inside a game so that we can
handle the even and something like that?

**Comment by itanasi** [comment]
Whatever you define as an event

**Pull request merged**